pace-util/pace/util/caching_comm.py

```diff
@@ -0,0 +1,201 @@
+import dataclasses
+import pickle
+from typing import Any, BinaryIO, List, Optional, TypeVar
+
+import numpy as np
+
+from .comm import Comm, Request
+
+
+T = TypeVar("T")
+
+
+class CachingRequestWriter(Request):
+    def __init__(self, req: Request, buffer: np.ndarray, buffer_list: List[np.ndarray]):
+        self._req = req
+        self._buffer = buffer
+        self._buffer_list = buffer_list
+
+    def wait(self):
+        self._req.wait()
+        self._buffer_list.append(copy.deepcopy(self._buffer))
+
+
+class CachingRequestReader(Request):
+    def __init__(self, recvbuf, data):
+        self._recvbuf = recvbuf
+        self._data = data
+
+    def wait(self):
+        self._recvbuf[:] = self._data
+
+
+class NullRequest(Request):
+    def wait(self):
+        pass
+
+
+@dataclasses.dataclass
+class CachingCommData:
+    """
+    Data required to restore a CachingCommReader.
+
+    Usually you will not want to initialize this class directly, but instead
+    use the CachingCommReader.load method.
+    """
+
+    rank: int
+    size: int
+    bcast_objects: List[Any] = dataclasses.field(default_factory=list)
+    received_buffers: List[np.ndarray] = dataclasses.field(default_factory=list)
+    split_data: List["CachingCommData"] = dataclasses.field(default_factory=list)
+
+    def __post_init__(self):
+        self._i_bcast = 0
+        self._i_buffers = 0
+        self._i_split = 0
+
+    def get_bcast(self):
+        return_value = self.bcast_objects[self._i_bcast]
+        self._i_bcast += 1
+        return return_value
+
+    def get_buffer(self):
+        return_value = self.received_buffers[self._i_buffers]
+        self._i_buffers += 1
+        return return_value
+
+    def get_split(self):
+        return_value = self.split_data[self._i_split]
+        self._i_split += 1
+        return return_value
+
+    def dump(self, file: BinaryIO):
+        pickle.dump(self, file)
+
+    @classmethod
+    def load(self, file: BinaryIO) -> "CachingCommData":
+        return pickle.load(file)
+
+
+class CachingCommReader(Comm):
+    """
+    mpi4py Comm-like object which replays stored communications.
+    """
+
+    def __init__(self, data: CachingCommData):
+        """
+        Initialize a CachingCommReader.
+
+        Usually you will not want to initialize this class directly, but instead
+        use the CachingCommReader.load method.
+
+        Args:
+            data: contains all data needed for mocked communication
+        """
+        self._data = data
+
+    def Get_rank(self) -> int:
+        return self._data.rank
+
+    def Get_size(self) -> int:
+        return self._data.size
+
+    def bcast(self, value: Optional[T], root=0) -> T:
+        return self._data.get_bcast()
+
+    def barrier(self):
+        pass
+
+    def Scatter(self, sendbuf, recvbuf, root=0, **kwargs):
+        recvbuf[:] = self._data.get_buffer()
+
+    def Gather(self, sendbuf, recvbuf, root=0, **kwargs):
+        recvbuf[:] = self._data.get_buffer()
```

Comment:
get_buffer will sometimes return None for this, because scatter/gather don't always receive data on every rank. Have to edit to fix this, should probably add a test.